source/opt/dead_branch_elim_pass.cpp

```diff
@@ -0,0 +1,335 @@
+        &condVal))
+      continue;
+
+    // Replace conditional branch with unconditional branch
+    const uint32_t trueLabId =
+        br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
+    const uint32_t falseLabId =
+        br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
+    const uint32_t mergeLabId =
+        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+    const uint32_t liveLabId = condVal == true ? trueLabId : falseLabId;
+    const uint32_t deadLabId = condVal == true ? falseLabId : trueLabId;
+    AddBranch(liveLabId, *bi);
+    def_use_mgr_->KillInst(br);
+    def_use_mgr_->KillInst(mergeInst);
+
+    // Iterate to merge block adding dead blocks to elimination set
+    std::unordered_set<uint32_t> deadSuccIds;
+    deadSuccIds.insert(deadLabId);
+    auto dbi = bi;
+    ++dbi;
+    uint32_t dLabId = (*dbi)->id();
+    while (dLabId != mergeLabId) {
+      if (deadSuccIds.find(dLabId) != deadSuccIds.end()) {
+        // Add successor blocks to dead successor set
+        (*dbi)->ForEachSuccessorLabel([&deadSuccIds](uint32_t succ) {
+          deadSuccIds.insert(succ);
+        });
+        // If dead block is merge block, add its merge block to dead
+        // successor set in case it has no predecessors.
+        uint32_t dummy;
+        const uint32_t dMergeLabId = MergeBlockIdIfAny(**dbi, &dummy);
+        if (dMergeLabId != 0)
+          deadSuccIds.insert(dMergeLabId);
+        // Kill use/def for all instructions and mark block for elimination
+        KillAllInsts(*dbi);
+        elimBlocks.insert(*dbi);
+      }
+      ++dbi;
+      dLabId = (*dbi)->id();
+    }
+
+    // Process phi instructions in merge block.
+    // deadLabIds are now blocks which cannot precede merge block.
+    // If eliminated branch is to merge label, add current block to dead blocks.
+    if (deadLabId == mergeLabId)
+      deadSuccIds.insert((*bi)->id());
+    (*dbi)->ForEachPhiInst([&deadSuccIds, this](ir::Instruction* phiInst) {
+      const uint32_t phiLabId0 =
+          phiInst->GetSingleWordInOperand(kPhiLab0IdInIdx);
+      const bool useFirst = deadSuccIds.find(phiLabId0) == deadSuccIds.end();
+      const uint32_t phiValIdx =
+          useFirst ? kPhiVal0IdInIdx : kPhiVal1IdInIdx;
+      const uint32_t replId = phiInst->GetSingleWordInOperand(phiValIdx);
+      const uint32_t phiId = phiInst->result_id();
+      (void)def_use_mgr_->ReplaceAllUsesWith(phiId, replId);
+      def_use_mgr_->KillInst(phiInst);
+    });
+
+    // If merge block has no predecessors, replace the new branch with
+    // a MergeSelection/BranchCondition using the original constant condition
+    // and the mergeblock as the false branch. This is done so the merge block
+    // is not orphaned, which could cause invalid control flow in certain case.
+    // TODO(greg-lunarg): Do this only in cases where invalid code is caused.
+    if (!HasNonPhiRef(mergeLabId)) {
```

Comment:
I had missed this point:

> At the point that I test if the (selection) merge block has non-phi refs, I have already removed the selection merge instruction that pointed to it. So we can get an unreachable (former) merge block. I even added a test that hits this very logic.

Looking again, I see the OrphanMerge, and experimentation proves that it triggers that case.

In reading the code, I got the fact that the loop finding dead blocks avoids adding the merge block, but then missed the significance of the "if (deadLabId == mergeLabId) deadSuccIds.insert(...) " code sequence.

Ok. Looks good to me now.  You've even got the TODO there.

Fair enough about the latch edge issue.  I'm more comfortable being aggressive there because I've already seen CTS test cases working even though they use the more aggressive rule. I only made the rule when I noticed the CTS test cases need something more general than the spec or validator was willing to give.